plugins/destination/batchingwriter/batchingwriter.go

```diff
@@ -0,0 +1,268 @@
+		metrics.Writes += atomic.LoadUint64(&m.Writes)
+	}
+	w.metricsLock.RUnlock()
+	return metrics
+}
+
+type worker struct {
+	count      int
+	wg         *sync.WaitGroup
+	ch         chan arrow.Record
+	flush      chan chan bool
+	sourceSpec specs.Source
+	openError  bool
+}
+
+func (*Batching) dummyWork(_ context.Context, ch <-chan arrow.Record, flush <-chan chan bool) {
+	for {
+		select {
+		case _, ok := <-ch:
+			if !ok {
+				return
+			}
+		case done := <-flush:
+			done <- true
+		}
+	}
+}
+func (w *Batching) work(ctx context.Context, sourceSpec specs.Source, syncTime time.Time, metrics *destination.Metrics, table *schema.Table, ch <-chan arrow.Record, flush <-chan chan bool) {
+	sizeBytes := int64(0)
+	resources := make([]arrow.Record, 0)
+	for {
+		select {
+		case r, ok := <-ch:
+			if !ok {
+				if len(resources) > 0 {
+					w.flush(ctx, sourceSpec, syncTime, metrics, table, resources)
+				}
+				return
+			}
+			if int64(len(resources)) == w.batchSize || sizeBytes+util.TotalRecordSize(r) > w.batchSizeBytes {
+				w.flush(ctx, sourceSpec, syncTime, metrics, table, resources)
+				resources = make([]arrow.Record, 0)
+				sizeBytes = 0
+			}
+			resources = append(resources, r)
+			sizeBytes += util.TotalRecordSize(r)
+		case <-time.After(w.batchTimeout):
+			if len(resources) > 0 {
+				w.flush(ctx, sourceSpec, syncTime, metrics, table, resources)
+				resources = make([]arrow.Record, 0)
+				sizeBytes = 0
+			}
+		case done := <-flush:
+			if len(resources) > 0 {
+				w.flush(ctx, sourceSpec, syncTime, metrics, table, resources)
+				resources = make([]arrow.Record, 0)
+				sizeBytes = 0
+			}
+			done <- true
+		}
+	}
+}
+
+func (w *Batching) flush(ctx context.Context, sourceSpec specs.Source, syncTime time.Time, metrics *destination.Metrics, table *schema.Table, resources []arrow.Record) {
+	if w.dedupPK {
+		resources = destination.RemoveDuplicatesByPK(table, resources)
+	}
+	start := time.Now()
+	batchSize := len(resources)
+
+	if err := w.underlyingWriter.WriteTableBatch(ctx, sourceSpec, table, syncTime, resources); err != nil {
+		w.logger.Err(err).Str("table", table.Name).Int("len", batchSize).Dur("duration", time.Since(start)).Msg("failed to write batch")
+		// we don't return an error as we need to continue until channel is closed otherwise there will be a deadlock
+		atomic.AddUint64(&metrics.Errors, uint64(batchSize))
+	} else {
+		w.logger.Info().Str("table", table.Name).Int("len", batchSize).Dur("duration", time.Since(start)).Msg("batch written successfully")
+		atomic.AddUint64(&metrics.Writes, uint64(batchSize))
+	}
+}
+
+func (w *Batching) Write(ctx context.Context, sourceSpec specs.Source, tables schema.Tables, syncTime time.Time, res <-chan arrow.Record) error {
+	workers := make(map[string]*worker, len(tables))
+
+	w.workersLock.Lock()
+	for _, table := range tables {
+		table := table
+		metrics := &destination.Metrics{}
+		w.metricsLock.Lock()
+		w.metrics[table.Name] = metrics
```

Comment:
This was broken/missing from main. Also we now do `atomic. LoadUint64` when tallying metrics.